cli/util/uploader.go

```diff
@@ -328,129 +343,211 @@ func updateIndexFiles(ctx *cli.Context, p *pool.Pool, containerID cid.ID, accoun
+	var (
+		res = make(chan oid.ID, 2*searchBatchSize)
+		wg  sync.WaitGroup
+	)
+
+	for j := int(startIndex); j < int(endIndex); j += searchBatchSize * maxParallelSearches {
+		for i := range maxParallelSearches {
+			start := j + i*searchBatchSize
+			end := start + searchBatchSize
+
+			if start >= int(endIndex) {
+				break
+			}
+			if end > int(endIndex) {
+				end = int(endIndex)
+			}
+
+			wg.Add(1)
+			go func(start, end int) {
+				defer wg.Done()
 
-				prm = client.PrmObjectSearch{}
-				filters = object.NewSearchFilters()
-				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+				prm := client.PrmObjectSearch{}
+				filters := object.NewSearchFilters()
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", start), object.MatchNumGE)
 				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLT)
 				prm.SetFilters(filters)
+
 				var objIDs []oid.ID
 				err := retry(func() error {
-					var errSearchIndex error
-					objIDs, errSearchIndex = neofs.ObjectSearch(ctx.Context, p, account.PrivateKey(), containerID.String(), prm)
-					return errSearchIndex
+					var errBlockSearch error
+					objIDs, errBlockSearch = neofs.ObjectSearch(ctx, p, account.PrivateKey(), containerID.String(), prm)
+					return errBlockSearch
 				})
-
 				if err != nil {
 					select {
-					case errCh <- fmt.Errorf("failed to search for objects from %d to %d for index file %d: %w", j, end, i, err):
+					case errCh <- fmt.Errorf("failed to search for block(s) from %d to %d: %w", start, end, err):
 					default:
 					}
 					return
 				}
 
 				for _, id := range objIDs {
-					oidCh <- id
-				}
-			}
-		}()
-
-		var completed int
-	waitLoop:
-		for {
-			select {
-			case err := <-errCh:
-				return err
-			case <-oidFetcherToProcessor:
-				completed++
-				if completed == int(indexFileSize) {
-					break waitLoop
+					res <- id
 				}
-			}
+			}(start, end)
 		}
-		// Check if there are any empty oids in the created index file.
-		// This could happen if object payload is empty ->
-		// attribute is not set correctly -> empty oid is added to the index file.
-		for k := 0; k < len(buffer); k += oidSize {
-			if slices.Compare(buffer[k:k+oidSize], emptyOid) == 0 {
-				return fmt.Errorf("empty oid found in index file %d at position %d (block index %d)", i, k/oidSize, i+uint(k/oidSize))
-			}
-		}
-		attrs := []object.Attribute{
-			*object.NewAttribute(attributeKey, strconv.Itoa(int(i))),
-			*object.NewAttribute("IndexSize", strconv.Itoa(int(indexFileSize))),
-		}
-		err := retry(func() error {
-			return uploadObj(ctx.Context, p, signer, account.PrivateKey().GetScriptHash(), containerID, buffer, attrs, homomorphicHashingDisabled)
-		})
-		if err != nil {
-			return fmt.Errorf("failed to upload index file %d: %w", i, err)
-		}
-		fmt.Fprintf(ctx.App.Writer, "Uploaded index file %d\n", i)
+		wg.Wait()
```

Comment:
Overall `searchObjects` looks good, but there's a problem: we can't wait here synchroniously. `searchObjects` must return almost immediately once it's called. Move the outer cycle to a separate routine. Move `wg` definition to the same routine. Remove the following code:
```
	go func() {
		wg.Wait()
		close(res)
	}()
```
Move `close(res)` to the same routine as the main cycle.